src/BenchmarkDotNet/Parameters/ParameterComparer.cs

```diff
@@ -18,30 +21,154 @@ public int Compare(ParameterInstances x, ParameterInstances y)
                 if (compareTo != 0)
                     return compareTo;
             }
+
             return string.CompareOrdinal(x.DisplayInfo, y.DisplayInfo);
         }
 
-        private int CompareValues(object x, object y)
+        private int CompareValues<T1, T2>(T1 x, T2 y)
         {
             // Detect IComparable implementations.
             // This works for all primitive types in addition to user types that implement IComparable.
-            if (x != null && y != null && x.GetType() == y.GetType() &&
-                x is IComparable xComparable)
+            if (x != null && y != null && x.GetType() == y.GetType())
+            {
+                if (x is IComparable xComparable)
+                {
+                    try
+                    {
+                        return xComparable.CompareTo(y);
+                    }
+                    // Some types, such as Tuple and ValueTuple, have a fallible CompareTo implementation which can throw if the inner items don't implement IComparable.
+                    // See: https://github.com/dotnet/BenchmarkDotNet/issues/2346
+                    // For now, catch and ignore the exception, and fallback to string comparison below.
+                    catch (ArgumentException ex) when (ex.Message.Contains("At least one object must implement IComparable."))
+                    {
+                    }
+                }
+                else if (x is IStructuralComparable)
+                {
+                    if (x is Array xArr && y is Array yArr)
+                    {
+                        if (xArr.Rank != yArr.Rank) return xArr.Rank.CompareTo(yArr.Rank);
+
+                        for (int dim = 0; dim < xArr.Rank; dim++)
+                        {
+                            if (xArr.GetLength(dim) != yArr.GetLength(dim))
+                                return xArr.GetLength(dim).CompareTo(yArr.GetLength(dim));
+                        }
+
+                        //  1D, 2D, and 3D array comparison is optimized with dedicated methods
+                        if (xArr.Rank == 1) return StructuralComparisonWithFallback(xArr, yArr);
+
+                        if (xArr.Rank == 2)
+                        {
+                            return (int) GetType()
+                                .GetMethod(nameof(CompareTwoDimensionalArray), BindingFlags.NonPublic | BindingFlags.Instance)
+                                .MakeGenericMethod(xArr.GetType().GetElementType(), yArr.GetType().GetElementType())
+                                .Invoke(this, [xArr, yArr]);
```

Comment:
We should guard these cases with `RuntimeInformation.IsAot` to prevent runtime exceptions from `MakeGenericMethod` in AOT runtimes.